### PR TITLE
Have streamly 0.7.3 build with ghc 9.4.

### PR DESCRIPTION
**Pull request by philderbeast**
Similar to #2167, this allows the `streamly-0.7.3.*` branch to be used with `ghc-9.4.5`.

### CODE REVIEW

File: src/Streamly/Internal/Data/List.hs [hunk 1]
```diff
@@ -6,6 +6,10 @@
 {-# LANGUAGE UndecidableInstances       #-} -- XXX
 {-# LANGUAGE ViewPatterns               #-}
 
+#if MIN_VERSION_base(4,17,0)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/List.hs:9]: Why is this required here?

File: src/Streamly/Internal/Data/Stream/Serial.hs [hunk 2]
```diff
@@ -8,6 +8,10 @@
 {-# LANGUAGE TypeFamilies              #-}
 {-# LANGUAGE UndecidableInstances      #-} -- XXX
 
+#if MIN_VERSION_base(4,17,0)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/Serial.hs:11]: Why is this required?

File: src/Streamly/Internal/Data/Stream/Zip.hs [hunk 3]
```diff
@@ -8,6 +8,10 @@
 {-# LANGUAGE TypeFamilies              #-}
 {-# LANGUAGE UndecidableInstances      #-} -- XXX
 
+#if MIN_VERSION_base(4,17,0)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/Zip.hs:11]: Why is this required?

File: streamly.cabal [hunk 4]
```diff
@@ -1,6 +1,6 @@
 cabal-version:      2.2
 name:               streamly
-version:            0.7.3
+version:            0.7.3.2
```
Comment (adithyaov) [inline streamly.cabal:3]: You should probably branch this from the latest 0.7.3.*

File: streamly.cabal [hunk 5]
```diff
@@ -289,9 +289,11 @@ library
...
                      , Streamly.Memory.Ring
 
-                     , Streamly.FileSystem.IOVec
-                     , Streamly.FileSystem.FDIO
-                     , Streamly.FileSystem.FD
+    if impl(ghc < 9)
+      other-modules:
+          Streamly.FileSystem.IOVec
+        , Streamly.FileSystem.FDIO
+        , Streamly.FileSystem.FD
```
Comment (adithyaov) [inline streamly.cabal:296]: Are these not supported post ghc 9?

**Comment by adithyaov** [comment]
@philderbeast

Sorry for the late review.
- Few comments
- Remove the cabal.project file
- Remove cabal.upgrade.project file
- Remove streamly.config file

Making CIs run for this will take some time.
You can build it locally test and push.


I'll check it post that from my end and make an upload dependingly.
I would recommend upgrading streamly to a newer version.

File: src/Streamly/Internal/Data/List.hs (hunk 1 above)
Comment (philderbeast) [reply src/Streamly/Internal/Data/List.hs:9]: To satisfy `-Wtype-equality-requires-operators`:

```
$ cabal build all --enable-tests --enable-benchmarks --project-file=cabal.upgrade.project
Build profile: -w ghc-9.4.5 -O1
...
[42 of 68] Compiling Streamly.Internal.Data.List

src/Streamly/Internal/Data/List.hs:117:13: warning: [-Wtype-equality-requires-operators]
    The use of ‘~’ without TypeOperators
    will become an error in a future GHC release.
    Suggested fix: Perhaps you intended to use TypeOperators
    |
117 | instance (a ~ Char) => IsString (List a) where
    |             ^

src/Streamly/Internal/Data/List.hs:179:13: warning: [-Wtype-equality-requires-operators]
    The use of ‘~’ without TypeOperators
    will become an error in a future GHC release.
    Suggested fix: Perhaps you intended to use TypeOperators
    |
179 | instance (a ~ Char) => IsString (ZipList a) where
    |             ^
 ```
 
 With earlier versions of GHC, this is not needed, hence the `CPP` conditional. The other option is not to `CPP` guard this language extension as `-XTypeOperators` has been available since `ghc-6.8.1`.

File: src/Streamly/Internal/Data/Stream/Zip.hs (hunk 3 above)
Comment (philderbeast) [reply src/Streamly/Internal/Data/Stream/Zip.hs:11]:
```
$ cabal build all --enable-tests --enable-benchmarks --with-compiler=ghc-9.4.5 \
  --ghc-options="-w -Wtype-equality-requires-operators"
Build profile: -w ghc-9.4.5 -O1
...
[37 of 68] Compiling Streamly.Internal.Data.Stream.Zip

src/Streamly/Internal/Data/Stream/Zip.hs:185:2923: warning: [-Wtype-equality-requires-operators]
    The use of ‘~’ without TypeOperators
    will become an error in a future GHC release.
    Suggested fix: Perhaps you intended to use TypeOperators
    |
185 | LIST_INSTANCES(ZipSerialM)
    |
 ```

File: src/Streamly/Internal/Data/List.hs (hunk 1 above)
Comment (philderbeast) [reply src/Streamly/Internal/Data/List.hs]: I removed the `CPP` guard but left the `-XTypeOperators` extension.

File: src/Streamly/Internal/Data/Stream/Serial.hs (hunk 2 above)
Comment (philderbeast) [reply src/Streamly/Internal/Data/Stream/Serial.hs]: I removed the `CPP` guard but left the `-XTypeOperators` extension.

File: src/Streamly/Internal/Data/Stream/Zip.hs (hunk 3 above)
Comment (philderbeast) [reply src/Streamly/Internal/Data/Stream/Zip.hs]: I removed the `CPP` guard but left the `-XTypeOperators` extension.

File: streamly.cabal (hunk 4 above)
Comment (philderbeast) [reply streamly.cabal:3]: Once #2421 merges, I'll be able to do this won't I?

File: streamly.cabal (hunk 5 above)
Comment (philderbeast) [reply streamly.cabal:296]: This is not a new change. It is included in #2421.

**Comment by philderbeast** [comment]
Hi @adithyaov, could you please have a look at #2421? If that is good to go and merged then I can change the merge target of this pull request.

**Comment by adithyaov** [comment]
@philderbeast I've changed the base to `branch-0.7.3.1` which is the latest `0.7.3.*`

File: streamly.cabal (hunk 4 above)
Comment (adithyaov) [reply streamly.cabal:3]: Resolved now. I've changed the base branch.

**Comment by adithyaov** [comment]
@philderbeast  Looks good.
Can you add back the CPP conditional? Adding `TypeOperators` for all GHCs without conditionals can break build or give unnecessary warnings on earlier GHCs.

**Comment by adithyaov** [comment]
Can you also update the GitHub workflow file adding GHC `9.2` and `9.4`?

**Comment by adithyaov** [comment]
Post that I can merge and upload a new version on hackage.

**Comment by philderbeast** [comment]
@adithyaov I've made the changes but you may need to approve the workflow as it is awaiting approval from a maintainer.

File: .github/workflows/haskell.yml [hunk 6]
```diff
@@ -44,39 +44,66 @@ jobs:
...
+          - name: 9.2.8
+            ghc_version: 9.2.8
+            build: cabal
+            cabal_version: 3.6.2.0
+            disable_sdist_build: "y"
+            ignore_error: false
+            runner: ubuntu-latest
           - name: 8.10.2+macOS
             ghc_version: 8.10.2
-            build: cabal-v2
```
Comment (adithyaov) [inline .github/workflows/haskell.yml:52]: This should be `cabal-v2`. `cabal` defaults to `cabal-v1` as this version of packcheck (The CI test tool) is old.

**Comment by adithyaov** [comment]
Approved the CIs to run.
One small comment for the CIs
I'll merge and upload the new version to hackage post that.

File: .github/workflows/haskell.yml (hunk 6 above)
Comment (philderbeast) [reply .github/workflows/haskell.yml:52]: I was comparing the command used in the latest version of master but I'll switch to `cabal-v2`.

**Comment by adithyaov** [comment]
Looks good @philderbeast 
Thank you for the work!
I'll merge the PR and upload a new version on Hackage.

**Comment by adithyaov** [comment]
I'll be squashing your commits into 2 logical commits post-merge.

**Comment by philderbeast** [comment]
@adithyaov I see there is one CI failure. Will you still be able to merge and publish to hackage?

**Comment by adithyaov** [comment]
@philderbeast Yes, I have been blocked on other things the past few days. 
I'll merge the PR and publish to hackage soon.
Sorry for the delay!

**Comment by adithyaov** [comment]
@philderbeast 0.7.3.2 has been published. 
See: https://hackage.haskell.org/package/streamly-0.7.3.2

**Comment by philderbeast** [comment]
Thanks @adithyaov. I've tried `streamly-0.7.3.2` and I need an `allow-newer: streamly:primitive` exception with it.

With https://github.com/space-vacuum/streamly/commit/59ef387efa32f4f7d192ef334f111ae269c90da8, I tried with `primitive-0.8.0.0` and the tests pass.

```
$ cabal test all --constraint="primitive ==0.8.0.0"
...
15 of 15 test suites (15 of 15 test cases) passed.
```

I see that `primitive-0.8.0.0` is distributed with stackage for `ghc-9.4.*` and with the latest nightlies.

https://www.stackage.org/package/primitive/snapshots

Would it be possible to make a hackage revision to `streamly-0.7.3.2` with this change?

```diff
 git diff
diff --git a/streamly.cabal b/streamly.cabal
index cdcfcaae..b681595d 100644
--- a/streamly.cabal
+++ b/streamly.cabal
@@ -386,7 +386,7 @@ library
                      , exceptions        >= 0.8   && < 0.11
                      , ghc-prim          >= 0.2   && < 0.10
                      , mtl               >= 2.2   && < 3
-                     , primitive         >= 0.5.4 && < 0.8
+                     , primitive         >= 0.5.4 && < 0.9
                      , transformers      >= 0.4   && < 0.6
 
                      , heaps             >= 0.3     && < 0.5
```

**Comment by adithyaov** [comment]
@philderbeast See https://github.com/composewell/streamly/pull/2469
I'll update the revision once the CIs pass.

**Comment by adithyaov** [comment]
@philderbeast 
Published a new revision.
The corresponding tag is `v0.7.3.2r1`

**Comment by philderbeast** [comment]
Works great, thank you. @adithyaov.
